### PR TITLE
Issue-258: add secureBatchReader

### PR DESCRIPTION
**Pull request by ravibeta**
### What This Change Is For
This change is for demonstrating batch read with tls configuration. This particular case was found to be tricky on a Kubernetes deployment because it involved setting up the PravegaCluster with K8s services and tls which introduces a relay. On the other hand, this creates a standalone program which makes it easier to use with the already popular secureWriter and secureReader.

Fixes https://github.com/pravega/pravega-samples/issues/258 
This comes from a request from test
Igor is ok with it.

### How This Works
This passes in a client certificate for tls connectivity and invokes the batch client

### How I Verified It Works
This was tested with a standalone mode

### Normal PR stuff:

- [] I've updated any docs that need to be updated
- [] I've added any unit tests that are easy to add
- [x] I've tested this works

```
Done creating a client config.
Done creating batchClient for myscope/mystream
Done reading event by thread 22: hello secure world!
Done reading 1 events
Done reading ranges of size: 1
All done with reading! Exiting...
```

### CODE REVIEW

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        try {
+            Futures.delayedTask(() -> {
+                logEventCounts();
+                this.executor.shutdownNow();
+                return null;
+            }, Duration.ofMinutes(0), executor).get();
+        } catch (InterruptedException | ExecutionException e) {
+           e.printStackTrace();
+        }
+        LOG.info("Started segment logger");
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:276]: Can we be consistent with logging? Here you use `LOG` but you also use `System.out.println` throughout this file.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    private final boolean validateHostName;
+
+    // Auth related config
+    private final String username;
+    private final String password;
+    private ClientConfig clientConfig;
+
+    private static Logger LOG = LoggerFactory.getLogger(SecureBatchReader.class);
+    private static final int THREAD_POOL_SIZE = 5;
+    private static ScheduledExecutorService streamCutLoggerService = newScheduledThreadPool(THREAD_POOL_SIZE, String.format("EventsCount"));
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:152]: Why static?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    private ScheduledExecutorService executor;
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:153]: This needs to be set in the constructor and you need a `close()` method on this class (implement `AutoCloseable`) to properly shut it down.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        CommandLineParser parser = new DefaultParser();
+        CommandLine cmd = parser.parse(options, args);
+        return cmd;
+    }
+
+    public void run() {
+        LOG.info("Starting segment logger");
+        this.executor = newScheduledThreadPool(THREAD_POOL_SIZE, String.format("EventsLogger"));
+        try {
+            Futures.delayedTask(() -> {
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:268]: Actually why do you use a future at all?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                logEventCounts();
+                this.executor.shutdownNow();
+                return null;
+            }, Duration.ofMinutes(0), executor).get();
+        } catch (InterruptedException | ExecutionException e) {
+           e.printStackTrace();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:274]: Let the exception bubble up.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 1 above)
Comment (andreipaduroiu) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:276]: Also, this message is misleading.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        }
+    }
+
+    private int readFromRanges(BatchClientFactory batchClient, List<SegmentRange> ranges) {
+        LOG.info("Number of ranges: {}", ranges.size());
+        List<CompletableFuture<Integer>> eventCounts = new ArrayList<>();
+        JavaSerializer<String> serializer = new JavaSerializer<String>();
+        eventCounts = ranges
+                .parallelStream()
+                .map(range -> CompletableFuture.supplyAsync(() -> batchClient.readSegment(range, serializer))
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:297]: You are executing this on the `ForkJoinPool.commonPool()` executor. Same with the callback. Switch it to use one of your executors.

Comment (andreipaduroiu) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:297]: Also, why do you execute the `readSegment` in a `supplyAsync` and then use `thenApply`? None of these are async; you can simplify your code by combining these in the same block of code.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+                                }
+                            } catch (Exception e) {
+                                e.printStackTrace();
+                            } finally {
+                                segmentIterator.close();
+                            }
+                            return numEvents;
+                        }))
+                .collect(Collectors.toList());
+        int count = eventCounts.stream().map(CompletableFuture::join).mapToInt(Integer::intValue).sum();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:315]: `Futures.allOfWithResults`

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+                            int numEvents = 0;
+                            try {
+                                String id = String.valueOf(Thread.currentThread().getId());
+                                while (segmentIterator.hasNext()) {
+                                    String event = segmentIterator.next();
+                                    System.out.println("Done reading event by thread " + id + ": " + event);
+                                    numEvents++;
+                                }
+                            } catch (Exception e) {
+                                e.printStackTrace();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:308]: Let the exception bubble up.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+            LOG.error("Exception:", e);
+        }
+    }
+
+    private int readFromRanges(BatchClientFactory batchClient, List<SegmentRange> ranges) {
+        LOG.info("Number of ranges: {}", ranges.size());
+        List<CompletableFuture<Integer>> eventCounts = new ArrayList<>();
+        JavaSerializer<String> serializer = new JavaSerializer<String>();
+        eventCounts = ranges
+                .parallelStream()
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:296]: There is no need for parallel stream here.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+                // TLS-related client-side configuration
+                .trustStore(this.truststorePath)
+                .validateHostName(this.validateHostName)
+
+                // Auth-related client-side configuration
+                .credentials(new DefaultCredentials(this.password, this.username))
+                .build();
+
+        System.out.println("Done creating a client config.");
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:202]: I think that we should be using `log` instead of `System.out` or `System.err` whenever possible. Can you please check if we use this kind of output normally in the rest of Pravega samples?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import java.util.Random;
+
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:110]: Nit: too many spaces.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+        }
+        LOG.info("Started segment logger");
+    }
+
+    private void logEventCounts() {
+        try (BatchClientFactory batchClient = BatchClientFactory.withScope(scope, clientConfig)) {
+             System.out.println("Done creating batchClient for " + scope + "/" + stream);
+             ArrayList<SegmentRange> ranges = Lists.newArrayList(batchClient.getSegments(Stream.of(scope, stream), StreamCut.UNBOUNDED, StreamCut.UNBOUNDED).getIterator());
+             readFromRanges(batchClient, ranges);
+             System.out.println("Done reading ranges of size: " + String.valueOf(ranges.size()));
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:284]: Same here about logging.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:286]: Use logging, but not both.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+        eventCounts = ranges
+                .parallelStream()
+                .map(range -> CompletableFuture.supplyAsync(() -> batchClient.readSegment(range, serializer))
+                        .thenApplyAsync(segmentIterator -> {
+                            int numEvents = 0;
+                            try {
+                                String id = String.valueOf(Thread.currentThread().getId());
+                                while (segmentIterator.hasNext()) {
+                                    String event = segmentIterator.next();
+                                    System.out.println("Done reading event by thread " + id + ": " + event);
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:304]: Again output to console.

**Comment by RaulGracia** [comment]
@ravibeta In samples, we have a different flow compared to other repositories: we target PRs to `dev` branch and commit there. Once we release, we release from `dev` and then merge `dev` into `master`, so it has the state of the last release. So please, retarget this PR to point to `dev` instead of `master`.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 1 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+
+    private static CommandLine parseCommandLineArgs(Options options, String[] args) throws ParseException {
+        CommandLineParser parser = new DefaultParser();
+        CommandLine cmd = parser.parse(options, args);
+        return cmd;
+    }
+
+    public void run() {
+        LOG.info("Starting segment logger");
+        this.executor = newScheduledThreadPool(THREAD_POOL_SIZE, String.format("EventsLogger"));
```
[the remark this answers is not in the document]
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 2 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre removed

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 3 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 5 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 6 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 7 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 8 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 9 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 10 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:125]: Yes Andre

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 11 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Raul

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 12 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Raul

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 13 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:194]: Yes Raul

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 14 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Raul

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+            System.out.println("Done creating batchClient for " + scope + "/" + stream);
+            ArrayList<SegmentRange> ranges = Lists.newArrayList(batchClient.getSegments(Stream.of(scope, stream), StreamCut.UNBOUNDED, StreamCut.UNBOUNDED).getIterator());
+            readFromRanges(batchClient, ranges);
+            System.out.println("Done reading ranges of size: " + String.valueOf(ranges.size()));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private int readFromRanges(BatchClientFactory batchClient, List<SegmentRange> ranges) {
```
Comment (ravisharda) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:198]: Since this code is for demo and samples, it should be as simple as it can be. Try to fold all of the major business logic in one method like is done here: 

https://github.com/pravega/pravega-samples/blob/ed9d0f883e07fcbc917c1ac7ce702ef3e31d84df/pravega-client-examples/src/main/java/io/pravega/example/secure/SecureReader.java#L72

It might help to think of someone who is totally new to Pravega and how he'd make sense of this code.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        List<CompletableFuture<Integer>> eventCounts = new ArrayList<>();
```
Comment (ravisharda) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:199]: The name of the variable doesn't reflect precisely what it is for. Can you give it a more specific name, especially for the benefit of Pravega users who might use this code?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+        JavaSerializer<String> serializer = new JavaSerializer<String>();
```
Comment (ravisharda) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:200]: Please simplify this code for the benefit of users. See a simpler example of batch reader here: 

https://github.com/pravega/pravega/blob/6b98ef07cbaeb997c930815f32002d59a76186ff/test/integration/src/test/java/io/pravega/test/integration/BatchClientTest.java#L219

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 16 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Ravi

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 17 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Ravi

File: pravega-client-examples/README.md [hunk 19]
```diff
@@ -152,7 +153,7 @@ First, ensure that the Pravega cluster that the applications are going to work w
 Auth (authentication and authorization) enabled. See Pravega documentation for how to enable TLS and Auth.
 
 Now, you might want to run `SecureWriter` in one window and `SecureReader` in another window.
-
+and finally `SecureBatchReader`
```
Comment (RaulGracia) [inline pravega-client-examples/README.md:156]: This looks disconnected from the preceding sentence, right? Maybe the dot should be a comma.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+
+        try (BatchClientFactory batchClient = BatchClientFactory.withScope(scope, clientConfig)) {
+            System.out.println("Done creating batchClient for " + scope + "/" + stream);
+            ArrayList<SegmentRange> segments = Lists.newArrayList(batchClient.getSegments(Stream.of(scope, stream), StreamCut.UNBOUNDED, StreamCut.UNBOUNDED).getIterator());
+            readFromSegments(batchClient, segments);
+            System.out.println("Done reading " + String.valueOf(segments.size()) + " segments.");
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        System.err.println("All done with reading! Exiting...");
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:138]: Why are you using `err`? is this an error or it can be printed by a successful execution?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+
+    public static void main(String[] args) throws ReinitializationRequiredException {
+        Options options = getOptions();
+
+        CommandLine cmd = null;
+        try {
+            cmd = parseCommandLineArgs(options, args);
+        } catch (ParseException e) {
+            System.out.format("%s.%n", e.getMessage());
+            final HelpFormatter formatter = new HelpFormatter();
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:149]: What is this for? Also `HelloWorldReader` does not seem related to this sample, which is a batch reader. Remove unnecessary stuff if possible to keep this simple.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+
+    private static CommandLine parseCommandLineArgs(Options options, String[] args) throws ParseException {
+        CommandLineParser parser = new DefaultParser();
+        CommandLine cmd = parser.parse(options, args);
+        return cmd;
+    }
+
+    private int readFromSegments(BatchClientFactory batchClient, List<SegmentRange> segments) {
+        List<Integer> batchEventCountList = new ArrayList<>();
+        UTF8StringSerializer serializer = new UTF8StringSerializer();
```
Comment (RaulGracia) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:196]: Are the other samples using this serializer, right? Otherwise it will have problems to read data from other samples.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/Constants.java [hunk 23]
```diff
@@ -16,7 +16,7 @@
 public class Constants {
     static final String DEFAULT_SCOPE = "myscope";
     static final String DEFAULT_STREAM_NAME = "mystream";
-    static final String DEFAULT_CONTROLLER_URI = "tls://localhost:9090";
+    static final String DEFAULT_CONTROLLER_URI = "tcp://10.247.78.154:9090";
```
Comment (ravisharda) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/Constants.java:19]: Revert

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+
+    // Auth related config
+    private final String username;
+    private final String password;
+    private ClientConfig clientConfig;
+
+    private static final int THREAD_POOL_SIZE = 5;
+    private ScheduledExecutorService batchCountExecutor;
+
+
```
Comment (ravisharda) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:73]: Remove this extra line

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+ * This class demonstrates how to configure a Pravega reader application for:
+ *
+ *    a) SSL/TLS (HTTPS) communications with a Pravega cluster for data-in-transit encryption and
+ *       server authentication.
+ *    b) Passing credentials to a Pravega cluster for authentication and authorization of clients.
+ *
+ * This example can be driven interactively against a running Pravega cluster configured to communicate using SSL/TLS
+ * and "auth" (authentication and authorization) turned on.
+ */
+public class SecureBatchReader implements AutoCloseable{
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:55]: Space before `{`

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 26]
```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2019 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:2]: Year

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    private final String truststorePath;
+    private final boolean validateHostName;
+
+    // Auth related config
+    private final String username;
+    private final String password;
+    private ClientConfig clientConfig;
+
+    private static final int THREAD_POOL_SIZE = 5;
+    private ScheduledExecutorService batchCountExecutor;
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:71]: Final

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+        System.out.println("Done creating a client config.");
+
+        // Everything below depicts the usual flow of reading events. All client-side security configuration is
+        // done through the ClientConfig object as shown above.
+
+        try (BatchClientFactory batchClient = BatchClientFactory.withScope(scope, clientConfig)) {
+            System.out.println("Done creating batchClient for " + scope + "/" + stream);
+            ArrayList<SegmentRange> segments = Lists.newArrayList(batchClient.getSegments(Stream.of(scope, stream), StreamCut.UNBOUNDED, StreamCut.UNBOUNDED).getIterator());
+            readFromSegments(batchClient, segments);
+            System.out.println("Done reading " + String.valueOf(segments.size()) + " segments.");
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:133]: Use `String.format` everywhere you need to output formatted strings. Do not use `String.valueOf`

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+                .build();
+
+        System.out.println("Done creating a client config.");
+
+        // Everything below depicts the usual flow of reading events. All client-side security configuration is
+        // done through the ClientConfig object as shown above.
+
+        try (BatchClientFactory batchClient = BatchClientFactory.withScope(scope, clientConfig)) {
+            System.out.println("Done creating batchClient for " + scope + "/" + stream);
+            ArrayList<SegmentRange> segments = Lists.newArrayList(batchClient.getSegments(Stream.of(scope, stream), StreamCut.UNBOUNDED, StreamCut.UNBOUNDED).getIterator());
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:131]: Use `ArrayList` constructor.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+            readFromSegments(batchClient, segments);
+            System.out.println("Done reading " + String.valueOf(segments.size()) + " segments.");
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:135]: So any exceptions here will still cause your program to exit successfully? This looks wrong.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    private static CommandLine parseCommandLineArgs(Options options, String[] args) throws ParseException {
+        CommandLineParser parser = new DefaultParser();
+        CommandLine cmd = parser.parse(options, args);
+        return cmd;
+    }
+
+    private int readFromSegments(BatchClientFactory batchClient, List<SegmentRange> segments) {
+        List<Integer> batchEventCountList = new ArrayList<>();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:192]: Why do you initialize this variable here? You are overwriting it 2 lines below.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+        JavaSerializer<String> serializer = new JavaSerializer<>();
+        batchEventCountList = segments
+                .stream()
+                .map(segment -> {
+                    SegmentIterator<String> segmentIterator = batchClient.readSegment(segment, serializer);
+                    int numEvents = 0;
+                    try {
+                        String id = String.valueOf(Thread.currentThread().getId());
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:200]: Do not use `String.valueOf`

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+                        while (segmentIterator.hasNext()) {
+                            String event = segmentIterator.next();
+                            System.out.println("Done reading event by thread " + id + ": " + event);
+                            numEvents++;
+                        }
+                    } catch (Exception e) {
+                        e.printStackTrace();
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:207]: So if you get an error you will return a partial result. This is incorrect. Bubble up the exception.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+                    } finally {
+                        segmentIterator.close();
+                    }
+                    return new Integer(numEvents);
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:211]: This looks weird. Why do you do wrap your int in an `Integer`?

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 35]
```diff
@@ -0,0 +1,218 @@
...
+                })
+                .collect(Collectors.toList());
+        int count = batchEventCountList.stream().mapToInt(Integer::intValue).sum();
+        System.out.println("Done reading " + String.valueOf(count) + " events");
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:215]: here too

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+                            numEvents++;
+                        }
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    } finally {
+                        segmentIterator.close();
+                    }
+                    return new Integer(numEvents);
+                })
+                .collect(Collectors.toList());
```
Comment (andreipaduroiu) [inline pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:213]: Just do `sum` here. Why do you create a list so you can just iterate through it?

Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 35 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 34 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 33 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:210]: Yes Andrei

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 32 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 29 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 28 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 27 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 26 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 25 above)
Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.

File: pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java (hunk 33 above)
Comment (andreipaduroiu) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java:210]: This is not done.

Comment (ravibeta) [reply pravega-client-examples/src/main/java/io/pravega/example/secure/SecureBatchReader.java]: Yes Andrei.
